keps/sig-node/20190226-migrate-in-tree-dockershim-to-out-of-tree.md

```diff
@@ -0,0 +1,148 @@
+---
+title: Separating a CRI for docker from Kubelet 
+authors:
+  - "@resouer"
+  - "@dims"
+owning-sig: sig-node
+reviewers:
+  - "@yujuhong"
+  - "@dchen1107"
+  - "@derekwaynecarr"
+  - "@PatrickLang"
+approvers:
+  - "@DawnChen"
+  - "@yujuhong"
+creation-date: 2019-02-26
+last-updated: 2019-02-26
+status: provisional
+---
+
+# Separating a CRI for Docker from Kubelet
+
+## Table of Contents
+
+- [Terms](#terms)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  * [Pros](#pros)
+  * [Cons](#cons)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  * [Dockershim deprecation plan](#dockershim-deprecation-plan)
+  * [Dockershim deprecation criteria](#dockershim-deprecation-criteria)
+  * [Test Plan](#test-plan)
+  * [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Terms
+
+- **CRI:** Container Runtime Interface – a plugin interface which enables kubelet to use a wide variety of container runtimes, without the need to recompile.
+
+
+## Summary
+
+CRI for docker (i.e. dockershim) is currently a build-in container runtime in kubelet code base. This proposal aims at a concrete deprecation and migration plan for separating dockershim from kubelet to out-of-tree without breaking current production users and WIP engineering efforts.
+
+## Motivation
+
+In Kubernetes, CRI is the used as the "default" container runtime, while currently the CRI of docker (a.k.a. dockershim) is part of kubelet code and coupled with kubelet's lifecycle. 
+
+This is not ideal as kubelet then has dependency on specific container runtime which leads to maintenance burden for not only developers in sig-node, but also cluster administrators when critical issues (e.g. runc CVE) happen to container runtimes. The pros of moving dockershim to out-of-tree is straightforward:
+
+### Pros
+- Docker is not special and just a CRI just like every other CRI in our ecosystem.
+- Currently, dockershim "enjoys" some backdoors for various reasons. Deprecating these "features" should eliminate maintenance burden of kubelet.
+- A cri-dockerd can be maintained independently.
+- Over time we can remove vendored docker dependencies in kubelet.
+
+Having said that, cons of deprecation build-in dockershim requires lots of attention:
+
+### Cons
+- Migration pain with a new binary in addition to kubelet.
+- The number of affected users maybe large.
+- Updating all the eco-system tools to support the new cri-dockerd.
```

Comment:
`Updating all the eco-system tools to support the new cri-dockerd binary to run alongside the kubelet`